File: boj/c++/2805.cpp
#include <iostream>
#define MAX 2000000000

using namespace std;

int trees[1000000];
int n, m, mid;
int answer;

int main(){
    scanf("%d %d", &n, &m);
    for (int i = 0; i <n; i++){
        scanf("%d", &trees[i]);
    }

    int start = 1, end = MAX;
    while (start <= end){
        mid = (start + end) / 2;
        long long int total = 0;
        for (int i = 0; i < n; i++)
            if (trees[i] > mid)
                total += trees[i] - mid;

        if (total < m)
            end = mid - 1;
        else{
            answer = mid;
            start = mid + 1;
        }
    }
    printf("%d", answer);
}

File: boj/python/2805.py
import sys


def input():
    return sys.stdin.readline()


sys.stdin = open('input.txt', 'r')


N, M = map(int, input().split())
l = list(map(int, input().split()))

start = 0
end = 1_000_000_000

while start <= end:
    mid = (start + end) // 2

    trees = 0
    for i in l:
        if i - mid > 0:
            trees += i - mid

    if trees >= M:
        start = mid + 1
    else:
        end = mid - 1

print(end)
